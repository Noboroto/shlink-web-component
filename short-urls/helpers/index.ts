import { isNil } from 'ramda';
import type { OptionalString } from '../../utils/helpers';
import type { ShortUrlCreationSettings } from '../../utils/settings';
import { DEFAULT_DOMAIN } from '../../visits/reducers/domainVisits';
import type { ShortUrl, ShortUrlData } from '../data';

export const shortUrlMatches = (shortUrl: ShortUrl, shortCode: string, domain: OptionalString): boolean => {
  if (isNil(domain)) {
    return shortUrl.shortCode === shortCode && !shortUrl.domain;
  }

  return shortUrl.shortCode === shortCode && shortUrl.domain === domain;
};

export const domainMatches = (shortUrl: ShortUrl, domain: string): boolean => {
  if (!shortUrl.domain && domain === DEFAULT_DOMAIN) {
    return true;
  }

  return shortUrl.domain === domain;
};

export const shortUrlDataFromShortUrl = (shortUrl?: ShortUrl, settings?: ShortUrlCreationSettings): ShortUrlData => {
  const validateUrl = settings?.validateUrls ?? false;

  if (!shortUrl) {
    return { longUrl: '', validateUrl };
  }

  return {
    longUrl: shortUrl.longUrl,
    tags: shortUrl.tags,
    title: shortUrl.title ?? undefined,
    domain: shortUrl.domain ?? undefined,
    validSince: shortUrl.meta.validSince ?? undefined,
    validUntil: shortUrl.meta.validUntil ?? undefined,
    maxVisits: shortUrl.meta.maxVisits ?? undefined,
    crawlable: shortUrl.crawlable,
    forwardQuery: shortUrl.forwardQuery,
    deviceLongUrls: shortUrl.deviceLongUrls,
    validateUrl,
  };
};

const MULTI_SEGMENT_SEPARATOR = '__';

export const urlEncodeShortCode = (shortCode: string): string => shortCode.replaceAll('/', MULTI_SEGMENT_SEPARATOR);

export const urlDecodeShortCode = (shortCode: string): string => shortCode.replaceAll(MULTI_SEGMENT_SEPARATOR, '/');
