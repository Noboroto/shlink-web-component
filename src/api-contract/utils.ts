import type {
  InvalidArgumentError,
  InvalidShortUrlDeletion,
  ProblemDetailsError,
} from '.';
import { ErrorType } from '.';

export const isInvalidArgumentError = (
  error?: ProblemDetailsError
): error is InvalidArgumentError => error?.type === ErrorType.INVALID_ARGUMENT;

export const isInvalidDeletionError = (
  error?: ProblemDetailsError
): error is InvalidShortUrlDeletion =>
  error?.type === ErrorType.INVALID_SHORT_URL_DELETION;

const isProblemDetails = (e: unknown): e is ProblemDetailsError =>
  !!e &&
  typeof e === 'object' &&
  ['type', 'detail', 'title', 'status'].every((prop) => prop in e);

export const parseApiError = (e: unknown): ProblemDetailsError | undefined =>
  isProblemDetails(e) ? e : undefined;

export const isErrorAction = (action: any): boolean => 'error' in action;
