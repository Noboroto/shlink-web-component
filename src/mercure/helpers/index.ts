import { EventSourcePolyfill as EventSource } from 'event-source-polyfill';
import type { MercureInfo } from '../reducers/mercureInfo';

export const bindToMercureTopic = <T>(
  mercureInfo: MercureInfo,
  topics: string[],
  onMessage: (message: T) => void,
  onTokenExpired: () => void
) => {
  const { mercureHubUrl, token, loading, error } = mercureInfo;

  if (loading || error || !mercureHubUrl) {
    return undefined;
  }

  const onEventSourceMessage = ({ data }: { data: string }) =>
    onMessage(JSON.parse(data) as T);
  const onEventSourceError = ({ status }: { status: number }) =>
    status === 401 && onTokenExpired();

  const subscriptions = topics.map((topic) => {
    const hubUrl = new URL(mercureHubUrl);

    hubUrl.searchParams.append('topic', topic);
    const es = new EventSource(hubUrl, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });

    es.onmessage = onEventSourceMessage;
    es.onerror = onEventSourceError;

    return es;
  });

  return () => subscriptions.forEach((es) => es.close());
};
