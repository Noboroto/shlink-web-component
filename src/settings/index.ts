export * from './types';
export * from './settingsContext';
