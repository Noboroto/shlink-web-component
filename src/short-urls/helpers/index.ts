import { isNil } from 'ramda';
import type { ShlinkShortUrl } from '../../api-contract';
import type { OptionalString } from '../../utils/helpers';
import type { ShortUrlCreationSettings } from '../../utils/settings';
import { DEFAULT_DOMAIN } from '../../visits/reducers/domainVisits';
import type { ShortUrlData } from '../data';

export const shortUrlMatches = (shortUrl: ShlinkShortUrl, shortCode: string, domain: OptionalString): boolean => {
  if (isNil(domain)) {
    return shortUrl.shortCode === shortCode && !shortUrl.domain;
  }

  return shortUrl.shortCode === shortCode && shortUrl.domain === domain;
};

export const domainMatches = (shortUrl: ShlinkShortUrl, domain: string): boolean => {
  if (!shortUrl.domain && domain === DEFAULT_DOMAIN) {
    return true;
  }

  return shortUrl.domain === domain;
};

export const shortUrlDataFromShortUrl = (
  shortUrl?: ShlinkShortUrl,
  settings?: ShortUrlCreationSettings,
): ShortUrlData => {
  const validateUrl = settings?.validateUrls ?? false;

  if (!shortUrl) {
    return { longUrl: '', validateUrl };
  }

  return {
    longUrl: shortUrl.longUrl,
    tags: shortUrl.tags,
    title: shortUrl.title ?? undefined,
    domain: shortUrl.domain ?? undefined,
    validSince: shortUrl.meta.validSince ?? undefined,
    validUntil: shortUrl.meta.validUntil ?? undefined,
    maxVisits: shortUrl.meta.maxVisits ?? undefined,
    crawlable: shortUrl.crawlable,
    forwardQuery: shortUrl.forwardQuery,
    deviceLongUrls: shortUrl.deviceLongUrls && {
      android: shortUrl.deviceLongUrls.android ?? undefined,
      ios: shortUrl.deviceLongUrls.ios ?? undefined,
      desktop: shortUrl.deviceLongUrls.desktop ?? undefined,
    },
    validateUrl,
  };
};

const MULTI_SEGMENT_SEPARATOR = '__';

export const urlEncodeShortCode = (shortCode: string): string => shortCode.replaceAll('/', MULTI_SEGMENT_SEPARATOR);

export const urlDecodeShortCode = (shortCode: string): string => shortCode.replaceAll(MULTI_SEGMENT_SEPARATOR, '/');
