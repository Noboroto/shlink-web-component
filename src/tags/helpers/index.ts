const ONE_OR_MORE_SPACES_REGEX = /\s+/g;

export const normalizeTag = (tag: string) => tag.trim().toLowerCase().replace(ONE_OR_MORE_SPACES_REGEX, '-');
