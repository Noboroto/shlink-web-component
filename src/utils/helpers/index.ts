import { isEmpty, isNil, pipe, range } from 'ramda';
import type { SyntheticEvent } from 'react';

export type OptionalString = string | null | undefined;

export const handleEventPreventingDefault = <T>(handler: () => T) => pipe(
  (e: SyntheticEvent) => e.preventDefault(),
  handler,
);

export const rangeOf = <T>(size: number, mappingFn: (value: number) => T, startAt = 1): T[] =>
  range(startAt, size + 1).map(mappingFn);

export type Empty = null | undefined | '' | never[];

export const hasValue = <T>(value: T | Empty): value is T => !isNil(value) && !isEmpty(value);

export const nonEmptyValueOrNull = <T>(value: T): T | null => (isEmpty(value) ? null : value);

export type BooleanString = 'true' | 'false';

export const parseBooleanToString = (value: boolean): BooleanString => (value ? 'true' : 'false');

export const parseOptionalBooleanToString = (value?: boolean): BooleanString | undefined => (
  value === undefined ? undefined : parseBooleanToString(value)
);
