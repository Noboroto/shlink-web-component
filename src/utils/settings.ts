import { createContext, useContext } from 'react';
import type { Theme } from '../../../src/utils/theme';
import type { ShortUrlsOrder } from '../short-urls/data';
import type { TagsOrder } from '../tags/data/TagsListChildrenProps';
import type { DateInterval } from './dates/helpers/dateIntervals';

export const DEFAULT_SHORT_URLS_ORDERING: ShortUrlsOrder = {
  field: 'dateCreated',
  dir: 'DESC',
};

/**
 * Important! When adding new props in the main Settings interface or any of the nested props, they have to be set as
 * optional, as old instances of the app will load partial objects from local storage until it is saved again.
 */

export interface RealTimeUpdatesSettings {
  enabled: boolean;
  interval?: number;
}

export type TagFilteringMode = 'startsWith' | 'includes';

export interface ShortUrlCreationSettings {
  validateUrls: boolean;
  tagFilteringMode?: TagFilteringMode;
  forwardQuery?: boolean;
}

export interface UiSettings {
  theme: Theme;
}

export interface VisitsSettings {
  defaultInterval: DateInterval;
  excludeBots?: boolean;
}

export interface TagsSettings {
  defaultOrdering?: TagsOrder;
}

export interface ShortUrlsListSettings {
  defaultOrdering?: ShortUrlsOrder;
}

export interface Settings {
  realTimeUpdates?: RealTimeUpdatesSettings;
  shortUrlCreation?: ShortUrlCreationSettings;
  shortUrlsList?: ShortUrlsListSettings;
  ui?: UiSettings;
  visits?: VisitsSettings;
  tags?: TagsSettings;
}

const defaultSettings: Settings = {
  realTimeUpdates: {
    enabled: true,
  },
  shortUrlCreation: {
    validateUrls: false,
  },
  ui: {
    theme: 'light',
  },
  visits: {
    defaultInterval: 'last30Days',
  },
  shortUrlsList: {
    defaultOrdering: DEFAULT_SHORT_URLS_ORDERING,
  },
};

const SettingsContext = createContext<Settings | undefined>(defaultSettings);

export const SettingsProvider = SettingsContext.Provider;

export const useSettings = (): Settings => useContext(SettingsContext) ?? defaultSettings;

export const useSetting = <T extends keyof Settings>(settingName: T): Settings[T] => {
  const settings = useSettings();
  return settings[settingName];
};
