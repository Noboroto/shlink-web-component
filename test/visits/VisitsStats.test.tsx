import { screen, waitFor } from '@testing-library/react';
import { fromPartial } from '@total-typescript/shoehorn';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import type { ShlinkVisit } from '../../src/api-contract';
import { rangeOf } from '../../src/utils/helpers';
import { SettingsProvider } from '../../src/utils/settings';
import type { VisitsInfo } from '../../src/visits/reducers/types';
import { VisitsStats } from '../../src/visits/VisitsStats';
import { checkAccessibility } from '../__helpers__/accessibility';
import { renderWithEvents } from '../__helpers__/setUpTest';

describe('<VisitsStats />', () => {
  const visits = rangeOf(3, () => fromPartial<ShlinkVisit>({ date: '2020-01-01' }));
  const getVisitsMock = vi.fn();
  const exportCsv = vi.fn();
  const setUp = (visitsInfo: Partial<VisitsInfo> = {}, activeRoute = '/by-time') => {
    const history = createMemoryHistory();
    history.push(activeRoute);

    return {
      history,
      ...renderWithEvents(
        <Router location={history.location} navigator={history}>
          <SettingsProvider value={fromPartial({})}>
            <VisitsStats
              getVisits={getVisitsMock}
              visitsInfo={fromPartial({ loading: false, error: false, loadingLarge: false, visits: [], ...visitsInfo })}
              cancelGetVisits={() => {}}
              exportCsv={exportCsv}
            />
          </SettingsProvider>
        </Router>,
      ),
    };
  };

  it('passes a11y checks', () => checkAccessibility(setUp()));

  it('renders a preloader when visits are loading', () => {
    setUp({ loading: true });

    expect(screen.getByText('Loading...')).toBeInTheDocument();
    expect(screen.queryByText(/^This is going to take a while/)).not.toBeInTheDocument();
  });

  it('renders a warning and progress bar when loading large amounts of visits', () => {
    setUp({ loading: true, loadingLarge: true, progress: 25 });

    expect(screen.queryByText('Loading...')).not.toBeInTheDocument();
    expect(screen.getByText(/^This is going to take a while/)).toBeInTheDocument();
    expect(screen.getByRole('progressbar')).toHaveAttribute('aria-valuenow', '25');
  });

  it('renders an error message when visits could not be loaded', () => {
    setUp({ error: true });
    expect(screen.getByText('An error occurred while loading visits :(')).toBeInTheDocument();
  });

  it('renders a message when visits are loaded but the list is empty', () => {
    setUp({ visits: [] });
    expect(screen.getByText('There are no visits matching current filter')).toBeInTheDocument();
  });

  it.each([
    ['/by-time', 2],
    ['/by-context', 4],
    ['/by-location', 3],
    ['/list', 1],
  ])('renders expected amount of charts', (route, expectedCharts) => {
    const { container } = setUp({ visits }, route);
    expect(container.querySelectorAll('.card')).toHaveLength(expectedCharts);
  });

  it('holds the map button on cities chart header', () => {
    setUp({ visits }, '/by-location');
    expect(
      screen.getAllByRole('img', { hidden: true }).some((icon) => icon.classList.contains('fa-map-location-dot')),
    ).toEqual(true);
  });

  it('exports CSV when export btn is clicked', async () => {
    const { user } = setUp({ visits });

    expect(exportCsv).not.toHaveBeenCalled();
    await user.click(screen.getByRole('button', { name: /Export/ }));
    expect(exportCsv).toHaveBeenCalled();
  });

  it('sets filters in query string', async () => {
    const { history, user } = setUp({ visits });
    const expectSearchContains = (contains: string[]) => {
      expect(contains).not.toHaveLength(0);
      contains.forEach((entry) => expect(history.location.search).toContain(entry));
    };

    expect(history.location.search).toEqual('');

    await user.click(screen.getByRole('button', { name: /Filters/ }));
    await waitFor(() => screen.getByRole('menu'));
    await user.click(screen.getByRole('menuitem', { name: 'Exclude potential bots' }));
    expectSearchContains(['excludeBots=true']);

    await user.click(screen.getByRole('button', { name: /Last 30 days/ }));
    await waitFor(() => screen.getByRole('menu'));
    await user.click(screen.getByRole('menuitem', { name: /Last 180 days/ }));
    expectSearchContains(['startDate', 'endDate']);
  });
});
