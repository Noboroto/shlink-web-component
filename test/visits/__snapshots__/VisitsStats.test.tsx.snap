// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<VisitsStats /> > loads visits when mounted 1`] = `
<div
  class="col-12 mt-3"
  data-testid="line-chart-container"
>
  <div
    class="card"
    data-testid="line-chart-card"
  >
    <div
      aria-level="4"
      class="d-flex justify-content-between align-items-center card-header"
      role="heading"
    >
      Visits over time
      <div
        class="d-flex align-content-center gap-1"
      >
        <button
          aria-controls=":r20:"
          aria-expanded="false"
          aria-label="Expand"
          class="btn btn-link btn-sm"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-up-right-and-down-left-from-center "
            data-icon="up-right-and-down-left-from-center"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M344 0H488c13.3 0 24 10.7 24 24V168c0 9.7-5.8 18.5-14.8 22.2s-19.3 1.7-26.2-5.2l-39-39-87 87c-9.4 9.4-24.6 9.4-33.9 0l-32-32c-9.4-9.4-9.4-24.6 0-33.9l87-87L327 41c-6.9-6.9-8.9-17.2-5.2-26.2S334.3 0 344 0zM168 512H24c-13.3 0-24-10.7-24-24V344c0-9.7 5.8-18.5 14.8-22.2s19.3-1.7 26.2 5.2l39 39 87-87c9.4-9.4 24.6-9.4 33.9 0l32 32c9.4 9.4 9.4 24.6 0 33.9l-87 87 39 39c6.9 6.9 8.9 17.2 5.2 26.2s-12.5 14.8-22.2 14.8z"
              fill="currentColor"
            />
          </svg>
        </button>
        <div
          class="d-flex align-items-center dropdown"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="btn-sm p-0 dropdown-toggle btn btn-link"
            type="button"
          >
            Group by
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu dropdown-menu-end"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item active"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Month
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Week
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Day
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Hour
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="card-body"
      id=":r20:"
    >
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 400px; min-width: 0;"
      />
      <ul
        class="list-unstyled mb-0 mt-2 d-flex flex-wrap justify-content-center gap-3"
      >
        <li
          class="d-inline"
        >
          <div
            class="color-bullet"
            style="background-color: rgb(70, 150, 229);"
          />
          <strong>
            Visits
             (
            3
            )
          </strong>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<VisitsStats /> > loads visits when mounted 2`] = `
<div
  class="col-12 mt-3"
  data-testid="line-chart-container"
>
  <div
    class="card"
    data-testid="line-chart-card"
  >
    <div
      aria-level="4"
      class="d-flex justify-content-between align-items-center card-header"
      role="heading"
    >
      Visits over time
      <div
        class="d-flex align-content-center gap-1"
      >
        <button
          aria-controls=":r23:"
          aria-expanded="false"
          aria-label="Expand"
          class="btn btn-link btn-sm"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-up-right-and-down-left-from-center "
            data-icon="up-right-and-down-left-from-center"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M344 0H488c13.3 0 24 10.7 24 24V168c0 9.7-5.8 18.5-14.8 22.2s-19.3 1.7-26.2-5.2l-39-39-87 87c-9.4 9.4-24.6 9.4-33.9 0l-32-32c-9.4-9.4-9.4-24.6 0-33.9l87-87L327 41c-6.9-6.9-8.9-17.2-5.2-26.2S334.3 0 344 0zM168 512H24c-13.3 0-24-10.7-24-24V344c0-9.7 5.8-18.5 14.8-22.2s19.3-1.7 26.2 5.2l39 39 87-87c9.4-9.4 24.6-9.4 33.9 0l32 32c9.4 9.4 9.4 24.6 0 33.9l-87 87 39 39c6.9 6.9 8.9 17.2 5.2 26.2s-12.5 14.8-22.2 14.8z"
              fill="currentColor"
            />
          </svg>
        </button>
        <div
          class="d-flex align-items-center dropdown"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="btn-sm p-0 dropdown-toggle btn btn-link"
            type="button"
          >
            Group by
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu dropdown-menu-end"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item active"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Month
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Week
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Day
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Hour
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="card-body"
      id=":r23:"
    >
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 400px; min-width: 0;"
      />
      <ul
        class="list-unstyled mb-0 mt-2 d-flex flex-wrap justify-content-center gap-3"
      >
        <li
          class="d-inline"
        >
          <div
            class="color-bullet"
            style="background-color: rgb(70, 150, 229);"
          />
          <strong>
            Visits
             (
            3
            )
          </strong>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<VisitsStats /> > loads visits when mounted 3`] = `
<div
  class="col-12 mt-3"
  data-testid="line-chart-container"
>
  <div
    class="card"
    data-testid="line-chart-card"
  >
    <div
      aria-level="4"
      class="d-flex justify-content-between align-items-center card-header"
      role="heading"
    >
      Visits over time
      <div
        class="d-flex align-content-center gap-1"
      >
        <button
          aria-controls=":r26:"
          aria-expanded="false"
          aria-label="Expand"
          class="btn btn-link btn-sm"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-up-right-and-down-left-from-center "
            data-icon="up-right-and-down-left-from-center"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M344 0H488c13.3 0 24 10.7 24 24V168c0 9.7-5.8 18.5-14.8 22.2s-19.3 1.7-26.2-5.2l-39-39-87 87c-9.4 9.4-24.6 9.4-33.9 0l-32-32c-9.4-9.4-9.4-24.6 0-33.9l87-87L327 41c-6.9-6.9-8.9-17.2-5.2-26.2S334.3 0 344 0zM168 512H24c-13.3 0-24-10.7-24-24V344c0-9.7 5.8-18.5 14.8-22.2s19.3-1.7 26.2 5.2l39 39 87-87c9.4-9.4 24.6-9.4 33.9 0l32 32c9.4 9.4 9.4 24.6 0 33.9l-87 87 39 39c6.9 6.9 8.9 17.2 5.2 26.2s-12.5 14.8-22.2 14.8z"
              fill="currentColor"
            />
          </svg>
        </button>
        <div
          class="d-flex align-items-center dropdown"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="btn-sm p-0 dropdown-toggle btn btn-link"
            type="button"
          >
            Group by
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu dropdown-menu-end"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item active"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Month
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Week
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Day
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Hour
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="card-body"
      id=":r26:"
    >
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 400px; min-width: 0;"
      />
      <ul
        class="list-unstyled mb-0 mt-2 d-flex flex-wrap justify-content-center gap-3"
      >
        <li
          class="d-inline"
        >
          <div
            class="color-bullet"
            style="background-color: rgb(70, 150, 229);"
          />
          <strong>
            Visits
             (
            3
            )
          </strong>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<VisitsStats /> > loads visits when mounted 4`] = `
<div
  class="col-12 mt-3"
  data-testid="line-chart-container"
>
  <div
    class="card"
    data-testid="line-chart-card"
  >
    <div
      aria-level="4"
      class="d-flex justify-content-between align-items-center card-header"
      role="heading"
    >
      Visits over time
      <div
        class="d-flex align-content-center gap-1"
      >
        <button
          aria-controls=":r29:"
          aria-expanded="false"
          aria-label="Expand"
          class="btn btn-link btn-sm"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-up-right-and-down-left-from-center "
            data-icon="up-right-and-down-left-from-center"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M344 0H488c13.3 0 24 10.7 24 24V168c0 9.7-5.8 18.5-14.8 22.2s-19.3 1.7-26.2-5.2l-39-39-87 87c-9.4 9.4-24.6 9.4-33.9 0l-32-32c-9.4-9.4-9.4-24.6 0-33.9l87-87L327 41c-6.9-6.9-8.9-17.2-5.2-26.2S334.3 0 344 0zM168 512H24c-13.3 0-24-10.7-24-24V344c0-9.7 5.8-18.5 14.8-22.2s19.3-1.7 26.2 5.2l39 39 87-87c9.4-9.4 24.6-9.4 33.9 0l32 32c9.4 9.4 9.4 24.6 0 33.9l-87 87 39 39c6.9 6.9 8.9 17.2 5.2 26.2s-12.5 14.8-22.2 14.8z"
              fill="currentColor"
            />
          </svg>
        </button>
        <div
          class="d-flex align-items-center dropdown"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="btn-sm p-0 dropdown-toggle btn btn-link"
            type="button"
          >
            Group by
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu dropdown-menu-end"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item active"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Month
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Week
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Day
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Hour
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="card-body"
      id=":r29:"
    >
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 400px; min-width: 0;"
      />
      <ul
        class="list-unstyled mb-0 mt-2 d-flex flex-wrap justify-content-center gap-3"
      >
        <li
          class="d-inline"
        >
          <div
            class="color-bullet"
            style="background-color: rgb(70, 150, 229);"
          />
          <strong>
            Visits
             (
            3
            )
          </strong>
        </li>
        <li
          class="d-inline"
        >
          <div
            class="color-bullet"
            style="background-color: rgb(29, 170, 88);"
          />
          <strong>
            Previous period
             (
            3
            )
          </strong>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<VisitsStats /> > loads visits when mounted 5`] = `
<div
  class="col-12 mt-3"
  data-testid="line-chart-container"
>
  <div
    class="card"
    data-testid="line-chart-card"
  >
    <div
      aria-level="4"
      class="d-flex justify-content-between align-items-center card-header"
      role="heading"
    >
      Visits over time
      <div
        class="d-flex align-content-center gap-1"
      >
        <button
          aria-controls=":r2c:"
          aria-expanded="false"
          aria-label="Expand"
          class="btn btn-link btn-sm"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-up-right-and-down-left-from-center "
            data-icon="up-right-and-down-left-from-center"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M344 0H488c13.3 0 24 10.7 24 24V168c0 9.7-5.8 18.5-14.8 22.2s-19.3 1.7-26.2-5.2l-39-39-87 87c-9.4 9.4-24.6 9.4-33.9 0l-32-32c-9.4-9.4-9.4-24.6 0-33.9l87-87L327 41c-6.9-6.9-8.9-17.2-5.2-26.2S334.3 0 344 0zM168 512H24c-13.3 0-24-10.7-24-24V344c0-9.7 5.8-18.5 14.8-22.2s19.3-1.7 26.2 5.2l39 39 87-87c9.4-9.4 24.6-9.4 33.9 0l32 32c9.4 9.4 9.4 24.6 0 33.9l-87 87 39 39c6.9 6.9 8.9 17.2 5.2 26.2s-12.5 14.8-22.2 14.8z"
              fill="currentColor"
            />
          </svg>
        </button>
        <div
          class="d-flex align-items-center dropdown"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="btn-sm p-0 dropdown-toggle btn btn-link"
            type="button"
          >
            Group by
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu dropdown-menu-end"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item active"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Month
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Week
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Day
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Hour
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="card-body"
      id=":r2c:"
    >
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 400px; min-width: 0;"
      />
      <ul
        class="list-unstyled mb-0 mt-2 d-flex flex-wrap justify-content-center gap-3"
      >
        <li
          class="d-inline"
        >
          <div
            class="color-bullet"
            style="background-color: rgb(70, 150, 229);"
          />
          <strong>
            Visits
             (
            3
            )
          </strong>
        </li>
        <li
          class="d-inline"
        >
          <div
            class="color-bullet"
            style="background-color: rgb(29, 170, 88);"
          />
          <strong>
            Previous period
             (
            3
            )
          </strong>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<VisitsStats /> > loads visits when mounted 6`] = `
<div
  class="col-12 mt-3"
  data-testid="line-chart-container"
>
  <div
    class="card"
    data-testid="line-chart-card"
  >
    <div
      aria-level="4"
      class="d-flex justify-content-between align-items-center card-header"
      role="heading"
    >
      Visits over time
      <div
        class="d-flex align-content-center gap-1"
      >
        <button
          aria-controls=":r2f:"
          aria-expanded="false"
          aria-label="Expand"
          class="btn btn-link btn-sm"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-up-right-and-down-left-from-center "
            data-icon="up-right-and-down-left-from-center"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M344 0H488c13.3 0 24 10.7 24 24V168c0 9.7-5.8 18.5-14.8 22.2s-19.3 1.7-26.2-5.2l-39-39-87 87c-9.4 9.4-24.6 9.4-33.9 0l-32-32c-9.4-9.4-9.4-24.6 0-33.9l87-87L327 41c-6.9-6.9-8.9-17.2-5.2-26.2S334.3 0 344 0zM168 512H24c-13.3 0-24-10.7-24-24V344c0-9.7 5.8-18.5 14.8-22.2s19.3-1.7 26.2 5.2l39 39 87-87c9.4-9.4 24.6-9.4 33.9 0l32 32c9.4 9.4 9.4 24.6 0 33.9l-87 87 39 39c6.9 6.9 8.9 17.2 5.2 26.2s-12.5 14.8-22.2 14.8z"
              fill="currentColor"
            />
          </svg>
        </button>
        <div
          class="d-flex align-items-center dropdown"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="btn-sm p-0 dropdown-toggle btn btn-link"
            type="button"
          >
            Group by
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu dropdown-menu-end"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item active"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Month
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Week
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Day
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Hour
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="card-body"
      id=":r2f:"
    >
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 400px; min-width: 0;"
      />
      <ul
        class="list-unstyled mb-0 mt-2 d-flex flex-wrap justify-content-center gap-3"
      >
        <li
          class="d-inline"
        >
          <div
            class="color-bullet"
            style="background-color: rgb(70, 150, 229);"
          />
          <strong>
            Visits
             (
            3
            )
          </strong>
        </li>
        <li
          class="d-inline"
        >
          <div
            class="color-bullet"
            style="background-color: rgb(29, 170, 88);"
          />
          <strong>
            Previous period
             (
            3
            )
          </strong>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
